### PR TITLE
Bug 1211561 - Destroy contexts when view is disposed of

### PR DESCRIPTION
**Pull request by stasm**
https://bugzilla.mozilla.org/show_bug.cgi?id=1211561

This reverts commit e12460a9da165d43b4e4750fd1c4abd654d9c72b:
"Flatten the cache and remove _resMap"

Conflicts:
	src/lib/context.js
	src/lib/env.js
	tests/lib/env/resCache.js

### CODE REVIEW

File: src/lib/env.js [hunk 1]
```diff
@@ -55,29 +67,44 @@ export class Env {
...
+    cache.set(id, resource);
+
+    return resource;
+  }
+}
+
+function deleteIfOrphan(cache, lists, resId) {
+  let isNeeded = false;
+  lists.forEach(
+    resIds => isNeeded = resIds.has(resId) ? true : isNeeded);
```
Comment (zbraniecki) [inline src/lib/env.js]: It would be possibly nice to stop looping on first `true`. Is there `Map` equivalent of `Array.some`?

Comment (stasm) [inline src/lib/env.js]: Unfortunately not.  I'm not a fan of `forEach` either and I haven't figured out how to use `Array.prototype.some.call` yet.

Comment (stasm) [inline src/lib/env.js]: I could perhaps use `Array.from(…).some(…)`.  @zbraniecki what's your take?

Comment (stasm) [inline src/lib/env.js]:
```javascript
const isNeeded = Array.from(lists).some(
  ([ctx, resIds]) => resIds.has(resId));
```

Comment (zbraniecki) [inline src/lib/env.js]: nah, leave it for now. Destructing will be a rare operation and the number of contexts probably below 10, so no need to over-optimize it for now imo.

Comment (stasm) [inline src/lib/env.js]: I agree it's going to be rare. Can I still use `some`? I think it tells a better story about this snippet's purpose.

Comment (zbraniecki) [inline src/lib/env.js]: sure, although it defies the purpose. The only value of some is if it allows you not to cycle through the whole list even once.
With your approach you loop twice in the conservative scenario.

**Pull request merged**